src/Common/src/IComparerHelpers.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace System
+{
+    internal static class IComparerHelpers
+    {
+        public static bool CompareReturnIfNull<T>([NotNullWhen(false)] T x, [NotNullWhen(false)] T y, [NotNullWhen(true)] out int? compareReturnValue)
```

Comment:
Don't you think that the fact that `compareReturnValue` is nullable is redundant with the fact that this method returns true? I guess you could either say that the contract of this method is that `compareReturnValue` shouldn't been use if the return value is false, a la `dictionary.TryGetValue`, or return a nullable int directly, right?

Reply:
I am not totally following you. I think either way you have to annotate it so the compiler knows the values aren't null later on.

Reply:
Whenever this method returns true, then `compareReturnValue` isn't null. I just think that its doesn't need to be an `int?` and could just be an `int`. Feel free to ignore if you disagree.

Reply:
The issue with that is 0 is a valid value for compare. So that is why I went for a null value. There isn't any way you can get it mixed up.

Reply:
That's true, but you could argue the same thing happen when you call `TryGetValue(key, out int value)` on a `Dictionary<T, int>`

Reply:
An alternative would be to have the signature of the method be
```suggestion
        public static int? CompareReturnIfNull<T>([NotNullWhen(false)] T x, [NotNullWhen(false)] T y)
```
as you current implementation effectively returns `compareReturnValue is null`.